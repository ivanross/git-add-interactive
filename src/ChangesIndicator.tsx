import React from 'react'
import { Box, Color } from 'ink'

interface Props {
  insertions?: number
  deletions?: number
}
export default function ChangesIndicator({ insertions, deletions }: Props) {
  if (!insertions && !deletions) return null
  return (
    <Box marginLeft={1}>
      (
      {insertions ? (
        <Box>
          +<Color green>{insertions}</Color>
        </Box>
      ) : null}
      {deletions ? (
        <Box>
          -<Color red>{deletions}</Color>
        </Box>
      ) : null}
      )
    </Box>
  )
}
