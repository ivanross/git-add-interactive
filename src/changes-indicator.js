const React = require("react");
const { Box, Color, useInput } = require("ink");

const ChangesIndicator = ({ insertions, deletions }) => {
  if (!insertions && !deletions) return null;
  return (
    <Box marginLeft={1}>
      (
      {insertions ? (
        <Box>
          +<Color green>{insertions}</Color>
        </Box>
      ) : null}
      {deletions ? (
        <Box>
          -<Color red>{deletions}</Color>
        </Box>
      ) : null}
      )
    </Box>
  );
};

module.exports = ChangesIndicator;
