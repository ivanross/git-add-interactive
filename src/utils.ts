import path from 'path'
// prettier-ignore
const statusMap = {
  "?": "untracked",
   M : "modified",
   D : "deleted",
   R : "renamed",
   A : "new_file",
};

// prettier-ignore
export const status2hex = {
  untracked: {"yellow": true},
  modified:  {"green":  true},
  deleted:   {"red":    true},
  renamed:   {"green":  true},
  new_file:  {"yellow": true}
};

export function cleanString(str: string) {
  let s = ''
  for (const c of str) {
    if (c === '"') continue
    if (c === ' ') s += '\\ '
    else s += c
  }
  return s
}

export function relativePath(from: string, to: string, filePath: string) {
  return path.relative(to, path.join(from, filePath))
}

export function getWorkingDirFilesInfo(files) {
  return files
    .filter(({ working_dir }) => statusMap[working_dir])
    .map(({ path, working_dir, insertions, deletions }) => ({
      path,
      status: statusMap[working_dir],
      insertions,
      deletions,
    }))
}

export function getIndexFilesInfo(files) {
  return files
    .filter(({ index, working_dir }) => statusMap[index] && working_dir !== '?')
    .map(({ path, index, insertions, deletions }) => ({
      path,
      status: statusMap[index],
      insertions,
      deletions,
    }))
}

export function parseFilesInfo(filesInfo, from?, to?) {
  return filesInfo
    .reduce((acc, { path, status, insertions, deletions }) => {
      const fromCWDFiles = path.split(' -> ').map(p => relativePath(from, to, p))
      const files = fromCWDFiles.map(cleanString)

      acc.push({
        files,
        status,
        label: fromCWDFiles.join(' -> '),
        value: fromCWDFiles.join(' -> '),
        insertions: insertions || 0,
        deletions: deletions || 0,
      })
      return acc
    }, [])
    .sort((a, b) => a.label.localeCompare(b.label))
}
