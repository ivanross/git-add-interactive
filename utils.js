"use strict";
const path = require("path");
// prettier-ignore
const statusMap = {
  "?": "untracked",
   M : "modified",
   D : "deleted",
   R : "renamed",
   A : "new_file",
};

// prettier-ignore
module.exports.status2hex = {
  untracked: {"yellow": true},
  modified:  {"green":  true},
  deleted:   {"red":    true},
  renamed:   {"green":  true},
  new_file:  {"yellow": true}
};

function cleanString(str) {
  let s = "";
  for (const c of str) {
    if (c === '"') continue;
    if (c === " ") s += "\\ ";
    else s += c;
  }
  return s;
}

module.exports.cleanString = cleanString;

function pathFromCWD(from, to) {
  return function(filePath) {
    return path.relative(to, path.join(from, filePath));
  };
}

module.exports.pathFromCWD = pathFromCWD;

module.exports.parseWorkingDirFiles = function parseWorkingDirFiles(
  files,
  parse = x => x
) {
  return files
    .reduce((acc, { path, working_dir }) => {
      const status = statusMap[working_dir];
      if (!status) return acc;

      const fromCwd = parse(path);
      const fileName = cleanString(fromCwd);
      acc.push({
        files: [fileName],
        status,
        value: fromCwd,
        label: fromCwd
      });
      return acc;
    }, [])
    .sort((a, b) => a.label.localeCompare(b.label));
};

module.exports.parseIndexFiles = function parseIndexFiles(
  files,
  parse = x => x
) {
  return files
    .reduce((acc, { path, index, working_dir }) => {
      const status = statusMap[index];
      if (!status || working_dir === "?") return acc;

      const fromCWDFiles = path.split(" -> ").map(parse);
      const files = fromCWDFiles.map(cleanString);

      acc.push({
        files,
        status,
        label: fromCWDFiles.join(" -> "),
        value: fromCWDFiles.join(" -> ")
      });
      return acc;
    }, [])
    .sort((a, b) => a.label.localeCompare(b.label));
};
