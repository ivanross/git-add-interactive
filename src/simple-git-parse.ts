import {
  FileStatusSumary,
  DiffResultTextFile,
  DiffResultBinaryFile,
} from 'simple-git/typings/response'
import { relativePath, cleanString, mergeBy } from './utils'

// prettier-ignore
const statusMap = {
  "?": "untracked",
  "M": "modified",
  "D": "deleted",
  "R": "renamed",
  "A": "new_file",
};

interface FileChangesInfo extends Partial<DiffResultTextFile & DiffResultBinaryFile> {}
export type FileInfo = ReturnType<typeof parse>[number]
export function parse(
  statusInfo: FileStatusSumary[],
  changesInfo: FileChangesInfo[],
  from: string,
  to: string,
  reset: boolean
) {
  const mergedInfo = mergeBy(statusInfo, changesInfo, ({ path }, { file }) => path === file)

  const filteredInfo = mergedInfo
    .filter(
      !reset
        ? ({ working_dir }) => statusMap[working_dir]
        : ({ index, working_dir }) => statusMap[index] && working_dir !== '?'
    )
    .map(info => {
      const status: string = statusMap[!reset ? info.working_dir : info.index]
      const fromCWDFiles = info.path.split(' -> ').map(p => relativePath(from, to, p))
      const files = fromCWDFiles.map(cleanString)
      const label = files.join(' -> ')

      return {
        status,
        files,
        label,
        value: label,
        insertions: info.insertions ?? 0,
        deletions: info.deletions ?? 0,
      }
    })

  return filteredInfo.sort((a, b) => a.label.localeCompare(b.label))
}
