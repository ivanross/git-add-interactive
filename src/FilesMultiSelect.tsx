import React from 'react'
import { Box, Color, Text, useInput } from 'ink'
import { status2hex } from './simple-git-parse'
import ChangesIndicator from './ChangesIndicator'

import figures from 'figures'
import logSymbols from 'log-symbols'
import MultiSelect from 'ink-multi-select'
import { FileInfo } from './simple-git-parse'
import { useList } from './useList'

const CheckBox = ({ isSelected }: { isSelected: boolean }) => (
  <Box marginRight={1}>
    <Text>{isSelected ? figures.circleFilled : figures.circle}</Text>
  </Box>
)

const Element = (w: number) => ({ label, status, insertions, deletions }: FileInfo) => {
  const color = status2hex[status]

  return (
    <Box>
      <Color {...color}>
        <Box width={w}>{status}:</Box>
        <Box>{label}</Box>
      </Color>
      <ChangesIndicator insertions={insertions} deletions={deletions} />
    </Box>
  )
}

const maxStatusWidth = (files: FileInfo[]) => {
  const arr = files.map(f => f.status.length)
  return Math.max(...arr)
}

interface Props {
  files: FileInfo[]
  onSubmit: (files: FileInfo[]) => void
  reset: boolean
}

export default function FilesMultiSelect({ files, onSubmit, reset }: Props) {
  const [selected, add, remove, set] = useList<FileInfo>([])

  const w = maxStatusWidth(files)

  useInput(input => {
    if (input === 'a' || input === 'A') {
      if (selected.length === files.length) set([])
      else set(files)
    }
  })

  return (
    <Box marginTop={1} flexDirection="column">
      <Box>
        <Box marginRight={1}>
          {logSymbols.info} Choose which file to {reset ? 'unstage' : 'stage'}.
        </Box>
        <Box>
          (Press <Color blue>{'<space>'}</Color> to select, <Color blue>{'<a>'}</Color> to toggle
          all)
        </Box>
      </Box>

      <Box>
        <Box marginLeft={4} width={w + 2}>
          <Text underline>status</Text>
        </Box>
        <Box marginLeft={4}>
          <Text underline>file</Text>
        </Box>
      </Box>

      <MultiSelect
        items={files}
        selected={selected}
        itemComponent={Element(w + 5)}
        checkboxComponent={CheckBox}
        onSubmit={onSubmit}
        onSelect={add}
        onUnselect={remove}
        limit={20}
      />
    </Box>
  )
}
