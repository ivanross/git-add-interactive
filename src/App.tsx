import React from 'react'
import FilesMultiSelect from './FilesMultiSelect'
import Summary from './Summary'
import { flatMap } from './utils'
import { SimpleGit } from 'simple-git/promise'
import { FileInfo } from './simple-git-parse'

interface Props {
  git: SimpleGit
  files: FileInfo[]
  reset: boolean
}

export default function App({ files, git, reset }: Props) {
  const [submitted, setSubmitted] = React.useState<FileInfo[]>([])

  const handleFilesSubmit = (items: FileInfo[]) => {
    if (items.length === 0) return

    const fileNames = flatMap(items, ({ files }) => files)

    if (reset) git.reset(fileNames).then(() => setSubmitted(items))
    else git.add(fileNames).then(() => setSubmitted(items))
  }

  return submitted.length === 0 ? (
    <FilesMultiSelect files={files} onSubmit={handleFilesSubmit} reset={reset} />
  ) : (
    <Summary files={submitted} reset={reset} />
  )
}
