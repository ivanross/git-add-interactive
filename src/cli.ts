#!/usr/bin/env node
import React from 'react'
import * as Ink from 'ink'
import meow from 'meow'
import Git, { SimpleGit } from 'simple-git/promise'
import App from './App'
import { parse } from './simple-git-parse'
import { DEFAULT_ERROR_MESSAGE } from './constants'

const cli = meow({
  help: `
  Usage
    
    $ git select-interactive [--reset]
    
  Select files to stage with interactive cli.
  Use arrows to navigate, <space> to select, <a> to 
  toggle all.

  Options
  
    --reset,   -r     Select files to unstage.
    --help,    -h     Show help
    --version, -v     Show version
`,
  autoHelp: true,
  flags: {
    help: {
      type: 'boolean',
      alias: 'h',
    },
    version: {
      type: 'boolean',
      alias: 'v',
    },
    reset: {
      type: 'boolean',
      alias: 'r',
    },
  },
})
const { reset } = cli.flags
const workingDir = process.cwd()

const git = Git(workingDir)

async function run(git: SimpleGit) {
  const isRepo = await git.checkIsRepo()
  if (!isRepo) {
    console.error(DEFAULT_ERROR_MESSAGE)
    process.exit(128)
  }

  const rootDir = await git.revparse(['--show-toplevel'])
  const changeInfo = (await git.diffSummary(reset ? ['--cached'] : [])).files
  const statusInfo = (await git.status()).files

  const files = parse(statusInfo, changeInfo, rootDir, workingDir, reset)
  if (files.length === 0) return

  Ink.render(React.createElement(App, { files, git, reset }))
}

run(git)
