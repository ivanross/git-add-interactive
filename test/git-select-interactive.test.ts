import { cleanString, relativePath, mergeBy } from '../src/utils'

//prettier-ignore
const test = [
  { path: "aa"               , index: " ", working_dir: "D" },
  { path: 'bb -> "dir/bb bb"', index: "R", working_dir: " " },
  { path: "new_file.txt"     , index: "A", working_dir: " " },
  { path: "test"             , index: "M", working_dir: " " },
  { path: "dir/cc"           , index: "?", working_dir: "?" },
  { path: "dir/test"         , index: "?", working_dir: "?" }
]

describe(cleanString, () => {
  it('should clean string', () => {
    expect(cleanString('test')).toBe('test')

    expect(cleanString('"a a"')).toBe('a\\ a')

    expect(cleanString('"src/foo/bar baz/f o o"')).toBe('src/foo/bar\\ baz/f\\ o\\ o')
  })
})

describe(relativePath, () => {
  it('should return path from current working dir', () => {
    expect(relativePath('/a/b/c', '/a/b/c/d', 'f')).toEqual('../f')
    expect(relativePath('/a/b/c', '/a/b', 'f')).toEqual('c/f')
  })
})

describe(mergeBy, () => {
  it('sohuld merge two array of objects', () => {
    const a = [
      { a: 1, b: 1 },
      { a: 2, b: 2 },
    ]
    const b = [
      { a: 1, c: 1 },
      { a: 3, c: 3 },
    ]
    expect(mergeBy(a, b, ({ a }, { c }) => a === c)).toEqual([
      { a: 1, b: 1, c: 1 },
      { a: 2, b: 2 },
    ])
  })
})
