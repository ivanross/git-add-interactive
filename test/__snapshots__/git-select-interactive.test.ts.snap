// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIndexFilesInfo should work 1`] = `
Array [
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "bb -> \\"dir/bb bb\\"",
    "status": "renamed",
  },
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "new_file.txt",
    "status": "new_file",
  },
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "test",
    "status": "modified",
  },
]
`;

exports[`getWorkingDirFilesInfo should work 1`] = `
Array [
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "aa",
    "status": "deleted",
  },
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "dir/cc",
    "status": "untracked",
  },
  Object {
    "deletions": undefined,
    "insertions": undefined,
    "path": "dir/test",
    "status": "untracked",
  },
]
`;

exports[`parseFilesInfo should work with index files 1`] = `
Array [
  Object {
    "deletions": 0,
    "files": Array [
      "bb",
      "dir/bb\\\\ bb",
    ],
    "insertions": 0,
    "label": "bb -> \\"dir/bb bb\\"",
    "status": "renamed",
    "value": "bb -> \\"dir/bb bb\\"",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "new_file.txt",
    ],
    "insertions": 0,
    "label": "new_file.txt",
    "status": "new_file",
    "value": "new_file.txt",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "test",
    ],
    "insertions": 0,
    "label": "test",
    "status": "modified",
    "value": "test",
  },
]
`;

exports[`parseFilesInfo should work with working dir files 1`] = `
Array [
  Object {
    "deletions": 0,
    "files": Array [
      "aa",
    ],
    "insertions": 0,
    "label": "aa",
    "status": "deleted",
    "value": "aa",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "dir/cc",
    ],
    "insertions": 0,
    "label": "dir/cc",
    "status": "untracked",
    "value": "dir/cc",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "dir/test",
    ],
    "insertions": 0,
    "label": "dir/test",
    "status": "untracked",
    "value": "dir/test",
  },
]
`;

exports[`parseFilesInfo should work with working dir files with different paths 1`] = `
Array [
  Object {
    "deletions": 0,
    "files": Array [
      "../../e/f/aa",
    ],
    "insertions": 0,
    "label": "../../e/f/aa",
    "status": "deleted",
    "value": "../../e/f/aa",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "../../e/f/dir/cc",
    ],
    "insertions": 0,
    "label": "../../e/f/dir/cc",
    "status": "untracked",
    "value": "../../e/f/dir/cc",
  },
  Object {
    "deletions": 0,
    "files": Array [
      "../../e/f/dir/test",
    ],
    "insertions": 0,
    "label": "../../e/f/dir/test",
    "status": "untracked",
    "value": "../../e/f/dir/test",
  },
]
`;
